AlphaWallet/Core/Coordinators/Alerts/AlertDataStoreType.swift

```diff
@@ -0,0 +1,144 @@
+//
+//  AlertDataStoreType.swift
+//  AlphaWallet
+//
+//  Created by Vladyslav Shepitko on 17.09.2021.
+//
+
+import UIKit
+import PromiseKit
+
+enum AlertUpdates {
+    case enabled(Bool)
+    case value(value: Double, marketPrice: Double)
+}
+
+protocol AlertDataStoreType: class {
+    var alertsSubscribable: Subscribable<[Alert]> { get }
+
+    func add(alert: Alert) -> Promise<Void>
+    func update(alert: Alert, update: AlertUpdates) -> Promise<Void>
+    func update(indexPath: IndexPath, update: AlertUpdates) -> Promise<Void>
+    func remove(indexPath: IndexPath) -> Promise<Void>
+}
+
+class AlertDataStore: AlertDataStoreType {
+    lazy var alertsSubscribable: Subscribable<[Alert]> = .init(load(forKey: key, defaultValue: []))
+    private let storage: StorageType
+    private enum Keys {
+        static func alertsKey(wallet: Wallet) -> String {
+            return "alerts-\(wallet.address.eip55String)"
+        }
+    }
+
+    private var alerts: [Alert] {
+        get {
+            if let value = alertsSubscribable.value {
+                return value
+            } else {
+                let value: [Alert] = load(forKey: key, defaultValue: [])
+                alertsSubscribable.value = value
+
+                return value
+            }
+        }
+        set {
+            guard let data = try? JSONEncoder().encode(newValue) else { return }
+
+            _ = storage.setData(data, forKey: key)
+            alertsSubscribable.value = newValue
+        }
+    }
+    private let key: String
+
+    init(storage: StorageType = FileStorage(), wallet: Wallet) {
+        self.storage = storage
+        self.key = AlertDataStore.Keys.alertsKey(wallet: wallet)
+    }
+
+    private func load<T: Codable>(forKey key: String, defaultValue: T) -> T {
+        guard let data = storage.data(forKey: key) else {
+            return defaultValue
+        }
+
+        guard let result = try? JSONDecoder().decode(T.self, from: data) else {
+            //NOTE: in case if decoding error appears, remove existed file
+            _ = storage.deleteEntry(forKey: key)
+
+            return defaultValue
+        }
+
+        return result
+    }
+
+    func add(alert: Alert) -> Promise<Void> {
+        return Promise { seal in
+            var alerts = self.alerts
+            alerts.append(alert)
```

Comment:
I think just `self.alerts.append(alert)` might work as expected because of how value types (the array) and observer works with value-type properties. i.e.

```
return Promise { seal in
    alerts.append(alert)
    seal.fulfill(())
}
```

Can you verify?

Reply:
checking